File: trees/find_only_children.py
class TreeNode:
  def __init__(self, value = 0, left = None, right = None):
    self.value = value
    self.left = left
    self.right = right



def findOnlyChildren(root):

  single_child_nodes = []

  def helper(root):
    if not root:
      return

    if not root.left is not None and root.right:
      single_child_nodes.append(root.right.value)

    elif root.left and not root.right is not None:
      single_child_nodes.append(root.left.value)
         
    # Traversing the left child
    helper(root.left)
     
    # Traversing the right child
    helper(root.right)
  
  helper(root)

  # print(single_child_nodes) 
  return single_child_nodes


print(findOnlyChildren(TreeNode(2, TreeNode(2), TreeNode(3, None, TreeNode(4)))) , [4])
print(findOnlyChildren(TreeNode(2, TreeNode(2, TreeNode(24), None), TreeNode(3, None, TreeNode(4)))), [24, 4])

File: binary_trees/findOnlyChildren.js
// this works in JS but not in Python b/c every time you call this function in Python, 
// b/c you have the array as the default [] its a mutable object and it will keep the old stuff from the old function call and mess everything up

function findOnlyChildren(root, result = []) {
  if (!root) return;

  if (root.right == null && root.left) {
    result.push(root.left.value);
  }
  if (root.left == null && root.right) {
    result.push(root.right.value);
  }
  findOnlyChildren(root.left, result);
  findOnlyChildren(root.right, result);
  return result;
}
